elephant/causality/granger.py

```diff
@@ -0,0 +1,438 @@
+    optimal_cov_mat : np.ndarray
+        covariance matrix of
+    optimal_order : int
+        optimal order
+    """
+
+    length = np.size(signals[0])
+
+    optimal_ic = np.infty
+    optimal_order = 1
+    optimal_coeffs = np.zeros((dimension, dimension, optimal_order))
+    optimal_cov_matrix = np.zeros((dimension, dimension))
+
+    for order in range(1, max_order + 1):
+        coeffs, cov_matrix = _vector_arm(signals, dimension, order)
+
+        temp_ic = information_criterion(cov_matrix, order, dimension, length)
+
+        if temp_ic < optimal_ic:
+            optimal_ic = temp_ic
+            optimal_order = order
+            optimal_coeffs = coeffs
+            optimal_cov_matrix = cov_matrix
+
+    return optimal_coeffs, optimal_cov_matrix, optimal_order
+
+
+def pairwise_granger(signals, max_order, information_criterion=aic):
+    r"""
+    Determine Granger Causality of two time series
+
+    Parameters
+    ----------
+    signals : np.ndarray or neo.AnalogSignal
+        time series data
+    max_order : int
+        maximal order of autoregressive model
+    information_criterion : callable
+        A function to compute the information criterion:
+            `bic` for Bayesian information_criterion,
+            `aic` for Akaike information criterion
+        Default: aic
+
+    Returns
+    -------
+    Causality
+        A `namedtuple` with the following attributes:
+            directional_causality_x_y : float
+                The Granger causality value for X influence onto Y.
+
+            directional_causality_y_x : float
+                The Granger causality value for Y influence onto X.
+
+            instantaneous_causality : float
+                The remaining channel interdependence not accounted for by
+                the directional causalities (e.g. shared input to X and Y).
+
+            total_interdependence : float
+                The sum of the former three metrics. It measures the dependence
+                of X and Y. If the total interdependence is positive, X and Y
+                are not independent.
+
+        Denote covariance matrix of signals
+            X by C|X  - a real number
+            Y by C|Y - a real number
+            (X,Y) by C|XY - a (2 \times 2) matrix
+
+        directional causality X -> Y given by
+            log(C|X / C|XY_00)
+        directional causality Y -> X given by
+            log(C|Y / C|XY_11)
+        instantaneous causality of X,Y given by
+            log(C|XY_00 / C|XY_11)
+        total interdependence of X,Y given by
+            log( {C|X \cdot C|Y} / det{C|XY} )
+
+    """
+    if isinstance(signals, AnalogSignal):
+        # transpose (N,2) -> (2, N)
+        signals = signals.magnitude.T
+    else:
+        signals = np.asarray(signals)
+
+    if not (signals.ndim == 2 and signals.shape[0] == 2):
+        raise ValueError("Required 2 x N input 'signals' array.")
+
+    # signal_x and signal_y are (1, N) arrays
```

Comment:
Not sure if this comment is required.

Reply:
The comment shows that the signals are not just vectors of length N as one might think.

Reply:
This might need a discussion. Perhaps we can avoid expanding dims in the pairwise and sort it out in the optimal vector arm?